File: python/680_Valid_Palindrome_II.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        if len(s) < 3:
            return True
        
        left = 0
        right = len(s) - 1
        
        while left < right:
            if s[left] != s[right]:
                return self.validSubPalindrome(s, left+1, right) or self.validSubPalindrome(s, left, right-1)
            
            left += 1
            right -= 1
                
        return True
    
    def validSubPalindrome(self, s: str, left: int, right: int) -> bool:        
        while left < right:
            if s[left] != s[right]:
                return False
            
            left += 1
            right -= 1
        
        return True

File: swift/680_valid_palindrome_II.swift
class Solution {
  func validPalindrome(_ s: String) -> Bool {
    let arrayS: [Character] = Array(s)

    var left: Int = 0
    var right: Int = arrayS.count - 1

    while left < right {
      if arrayS[left] != arrayS[right] {
        return isPalindrome(Array(arrayS[left + 1...right]))
          || isPalindrome(Array(arrayS[left..<right]))
        //                 let leftSkippedArray = Array(arrayS[left+1...right])
        //                 let rightSkippedArray = Array(arrayS[left..<right])

        //                 return (leftSkippedArray == leftSkippedArray.reversed()) || (rightSkippedArray == rightSkippedArray.reversed())
      }

      left += 1
      right -= 1
    }

    return true
  }

  private func isPalindrome(_ arrayS: [Character]) -> Bool {
    var left: Int = 0
    var right: Int = arrayS.count - 1

    while left < right {
      if arrayS[left] != arrayS[right] {
        return false
      }

      left += 1
      right -= 1
    }

    return true
  }
}
